### PR TITLE
Generate Yul SSA CFG

### PR DESCRIPTION
**Pull request by r0qs**
This PR implements a direct translation from Yul AST into a CFG in the SSA form. The implementation is based on the following paper: https://c9x.me/compile/bib/braun13cc.pdf

### CODE REVIEW

File: libyul/backends/evm/SSAControlFlowGraph.h [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+		struct FunctionReturn
+		{
+			langutil::DebugData::ConstPtr debugData;
+			std::vector<ValueId> returnValues;
+		};
+		struct Terminated {};
+		langutil::DebugData::ConstPtr debugData;
+		std::set<BlockId> entries;
+		std::set<ValueId> phis;
+		std::list<Operation> operations;
```
Comment (r0qs) [inline libyul/backends/evm/SSAControlFlowGraph.h]:
```suggestion
		std::vector<Operation> operations;
```

File: libyul/backends/evm/SSAControlFlowGraph.h [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	struct FunctionInfo {
+		langutil::DebugData::ConstPtr debugData;
+		std::reference_wrapper<Scope::Function const> function;
+		BlockId entry;
+		std::set<BlockId> exits;
+		bool canContinue = true;
+		std::vector<std::tuple<std::reference_wrapper<Scope::Variable const>, ValueId>> arguments;
+		std::vector<std::reference_wrapper<Scope::Variable const>> returns;
+	};
+	std::vector<std::reference_wrapper<Scope::Function const>> functions;
```
Comment (r0qs) [inline libyul/backends/evm/SSAControlFlowGraph.h:221]: We should have subgraphs for functions and the main entry graph

File: libyul/backends/evm/SSAControlFlowGraph.h [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+		m_blocks.emplace_back(BasicBlock{std::move(_debugData), {}, {}, {}, BasicBlock::Terminated{}});
+		return blockId;
+	}
+	BasicBlock& block(BlockId _id) { return m_blocks.at(_id.value); }
+	BasicBlock const& block(BlockId _id) const { return m_blocks.at(_id.value); }
+	size_t numBlocks() const { return m_blocks.size(); }
+private:
+	std::vector<BasicBlock> m_blocks;
+public:
+	struct LiteralValue { u256 value; };
```
Comment (r0qs) [inline libyul/backends/evm/SSAControlFlowGraph.h]: Add debugData to all values

File: libyul/backends/evm/SSAControlFlowGraphBuilder.cpp [hunk 4]
```diff
@@ -0,0 +1,736 @@
...
+		// TODO: also generate a subtraction to shift tight, but non-zero switch cases - or, alternative,
+		// transform to zero-based tight switches on Yul if possible.
+		std::map<u256, SSACFG::BlockId> cases;
+		std::optional<SSACFG::BlockId> defaultCase;
+		std::vector<std::tuple<SSACFG::BlockId, std::reference_wrapper<Block const>>> children;
+		for (auto const& _case: _switch.cases)
+		{
+			auto blockId = m_graph.makeBlock(debugDataOf(_case.body));
+			if (_case.value)
+				cases[_case.value->value.value()] = blockId;
```
Comment (r0qs) [inline libyul/backends/evm/SSAControlFlowGraphBuilder.cpp:370]: Cherry-pick the default switch case

File: libyul/backends/evm/SSAControlFlowGraph.h [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+			langutil::DebugData::ConstPtr debugData;
+			BlockId target;
+		};
+		struct JumpTable
+		{
+			langutil::DebugData::ConstPtr debugData;
+			ValueId value;
+			std::map<u256, BlockId> cases;
+			BlockId defaultCase;
+
```
Comment (matheusaaguiar) [inline libyul/backends/evm/SSAControlFlowGraph.h]:
```suggestion
```

File: libyul/backends/evm/SSAControlFlowGraph.h [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+			{
+				_callable(conditionalJump->nonZero);
+				_callable(conditionalJump->zero);
+			}
+			else if (auto* jumpTable = std::get_if<JumpTable>(&exit))
+			{
+				for (auto _case: jumpTable->cases | ranges::views::values)
+					_callable(_case);
+				_callable(jumpTable->defaultCase);
+			}
```
Comment (matheusaaguiar) [inline libyul/backends/evm/SSAControlFlowGraph.h:134]: The other types are just ignored?

File: libyul/backends/evm/SSAControlFlowGraph.h [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+		if (!m_unreachableValue)
+		{
+			m_unreachableValue = ValueId { m_valueInfos.size() };
+			m_valueInfos.emplace_back(UnreachableValue{});
+		}
+		return *m_unreachableValue;
+	}
+	ValueId newLiteral(langutil::DebugData::ConstPtr _debugData, u256 _value)
+	{
+		auto [it, inserted] = m_literals.emplace(std::make_pair(_value, ValueId{m_valueInfos.size()}));
```
Comment (matheusaaguiar) [inline libyul/backends/evm/SSAControlFlowGraph.h]:
```suggestion
		auto [it, inserted] = m_literals.emplace(_value, ValueId{m_valueInfos.size()});
```
I think `emplace` should be able to perfectly forward the arguments.

**Review by matheusaaguiar — COMMENTED** [review]
I know it's a draft, but I got curious and started looking at the code and reading the paper yesterday :sweat_smile:

File: libyul/backends/evm/SSAControlFlowGraph.h (hunk 7 above)
Comment (r0qs) [reply libyul/backends/evm/SSAControlFlowGraph.h]: Yes, we changed that in another branch haha

File: libyul/backends/evm/SSAControlFlowGraph.h (hunk 6 above)
Comment (clonker) [reply libyul/backends/evm/SSAControlFlowGraph.h:134]: For now yes, this is just for traversal of an individual graph (one graph per function and one main graph)

**Review by ekpyron — APPROVED** [review]
I've not reviewed this properly, but this is not critical for now and (together with the next PR) will be tested by the GreY project, while otherwise still not being public interface, so I'm approving and we can merge and then refine later when we move to actual codegen on top of it to properly test it.

**Pull request merged**